homeassistant/components/mqtt/fan.py

```diff
@@ -170,11 +222,38 @@ def _setup_from_config(self, config):
             "SPEED_HIGH": config[CONF_PAYLOAD_HIGH_SPEED],
             "SPEED_OFF": config[CONF_PAYLOAD_OFF_SPEED],
         }
+        if not self._topic[CONF_SPEED_COMMAND_TOPIC] is None:
+            self._feature_speeds = 1
+            _LOGGER.warning(
+                "Using speeds is deprecated, use preset modes or percentage instead"
+            )
+        else:
+            self._feature_speeds = 0
+
+        if not self._topic[CONF_PERCENTAGE_COMMAND_TOPIC] is None:
+            self._feature_percentage = 1
+        else:
+            self._feature_percentage = 0
+
+        if not self._topic[CONF_PRESET_MODE_COMMAND_TOPIC] is None:
+            self._feature_preset_mode = 1
+        else:
+            self._feature_preset_mode = 0
+        if not config[CONF_PRESET_MODES_LIST] and self._feature_preset_mode:
+            self._feature_preset_mode = 0
+            raise ValueError("No preset_modes configured, preset mode feature disabled")
```

Comment:
Don't raise here, add a check in the schema instead, something like:
```py
        vol.Inclusive(CONF_PRESET_MODE_COMMAND_TOPIC, "preset_mode"): mqtt.valid_publish_topic,
        vol.Inclusive(CONF_PRESET_MODES_LIST, "preset_mode"): cv.ensure_list,
```

Reply:
I updated the code and removed the raise.  I wil only check if an empty list was supplied. In that scenario preset mode set command will not effect state.

Reply:
Code was updated to resolve this issue.

Reply:
What I meant was to enforce that `CONF_PRESET_MODES_LIST` is provided together with `CONF_PRESET_MODE_COMMAND_TOPIC` already in the schema instead of waiting until the entity is setup.

Reply:
Ok, I will have al look how I can enforce this, any hint how toe do this without throwing an exception?

Reply:
I think I might have misled you earlier.
It's perfectly fine (and expected) for the schema to throw if the config is invalid, the problem before was that you raised _after_ the configuration had already been validated and the entity was setup.

Reply:
I have moved the checks as suggested. Further I removed checks that were not need after the entity is set up. Also processed Nicks comments.